pkg/rc/manager.go

```diff
@@ -0,0 +1,183 @@
+				// TODO: restart the watch so that you get updates right away?
+			}
+		case err := <-rcErr:
+			rcm.logger.WithError(err).Errorln("Could not read consul replication controllers")
+		case rcFields := <-rcWatch:
+			rcm.logger.WithField("n", len(rcFields)).Debugln("Received replication controller update")
+			if session == "" {
+				// we can't claim new nodes because our session is invalidated.
+				// raise an error and ignore this update
+				rcm.logger.NoFields().Infoln("Received replication controller update, but do not have session to acquire locks")
+				continue
+			}
+
+			// track which children were found in the returned set
+			foundChildren := make(map[fields.ID]struct{})
+			for _, rcField := range rcFields {
+				if _, ok := rcm.children[rcField.ID]; ok {
+					// this one is already ours, skip
+					rcm.logger.WithField("rc_id", rcField.ID).Debugln("Got replication controller already owned by self")
+					foundChildren[rcField.ID] = struct{}{}
+					continue
+				}
+
+				success, err := rcm.rcStore.Lock(rcField.ID, session)
+				if err != nil {
+					// log and break out of the loop - we can't continue if
+					// we had a real failure, our session probably disappeared
+					rcm.logger.WithErrorAndFields(err, logrus.Fields{
+						"rc_id":   rcField.ID,
+						"session": session,
+					}).Errorln("Got error while locking replication controller - session may be expired")
+					break // TODO: should we be pruning after this?
+				} else if !success {
+					// someone else must have gotten it first - log and move to
+					// the next one
+					rcm.logger.WithFields(logrus.Fields{
+						"rc_id":   rcField.ID,
+						"session": session,
+					}).Debugln("Lock on replication controller was denied")
+					continue
+				}
+
+				// at this point the rc is ours, time to spin it up
+				rcm.logger.WithFields(logrus.Fields{
+					"rc_id":   rcField.ID,
+					"session": session,
+				}).Infoln("Acquired lock on new replication controller, spawning")
+
+				childRC := New(rcField, rcm.kpStore, rcm.rcStore, rcm.scheduler, rcm.labeler)
+				childQuit := make(chan struct{})
+				rcm.children[rcField.ID] = managedRC{rc: childRC, quit: childQuit}
+				foundChildren[rcField.ID] = struct{}{}
+
+				go func() {
+					// disabled-ness is handled in watchdesires
+					for range childRC.WatchDesires(childQuit) {
+						rcm.logger.WithErrorAndFields(err, logrus.Fields{
+							"rc_id": childRC.ID(),
+						}).Errorln("Got error in replication controller loop")
+					}
+				}()
+			}
+
+			// now remove any children that were not found in the result set
+			rcm.logger.NoFields().Debugln("Pruning replication controllers that have disappeared")
+			for id := range rcm.children {
+				if _, ok := foundChildren[id]; !ok {
+					rcm.releaseChild(id)
+				}
+			}
+		}
+	}
+}
+
+// close one child
+func (rcm *RCManager) releaseChild(id fields.ID) {
+	rcm.logger.WithField("rc_id", id).Infoln("Releasing replication controller")
+	close(rcm.children[id].quit)
```

Comment:
that's a valid question in general, but i think the answer is:

- channels used primarily for quit should almost never be sensitive to the value being sent (and in our code they're all `struct{}` so by definition there is only one value that can be sent)
- range loop is terminated by close
- channels that are used in both directions should be very rare, and very obviously named